scheduledJobs/giveApples.js

```diff
@@ -0,0 +1,332 @@
+                }
+                filteredMembers.forEach((m) => membersNeedingOnlyRedApples.add(m));
+            } else {
+                // i.e. homework logbook
+                setLogbookUserIds(message, userIds);
+                filteredMembers.forEach((m) => membersNeedingApples.push(m));
+            }
+        });
+    };
+
+    const removeRollCallRole = async (member) => {
+        // roll call role removed from students given a '🍏'
+        const hasRollCollRole = member.roles.cache.has(rollCallRoleId);
+        const hasGreenAppleRole = member.roles.cache.has(greenAppleRoleId) || usersWithGreenAppleAdded.has(member.id);
+        if (hasRollCollRole && hasGreenAppleRole) {
+            await member.roles.remove([rollCallRoleId]);
+            usersWithRollCallRemoved.add(member.id);
+        }
+    };
+
+    const addActiveStudentRole = async (member) => {
+        // active student role added to students given a '🍎'
+        const hasActiveStudentRole = member.roles.cache.has(activeStudentRoleId);
+        const hasRedAppleRole = member.roles.cache.has(redAppleRoleId) || usersWithRedAppleAdded.has(member.id);
+        if (!hasActiveStudentRole && hasRedAppleRole) {
+            await member.roles.add([activeStudentRoleId]);
+            usersWithActiveStudentAdded.add(member.id);
+        }
+    };
+
+    const addAppleRoles = async () => {
+        const bothRoles = [greenAppleRoleId, redAppleRoleId];
+        if (logbookUserIds.length === 0 || membersNeedingApples.length === 0) {
+            return;
+        }
+
+        const userFrequency = {};
+        for (let i = 0; i < logbookUserIds.length; i++) {
+            userFrequency[logbookUserIds[i]] = userFrequency[logbookUserIds[i]] ? userFrequency[logbookUserIds[i]] + 1 : 1;
+        }
+
+        await Promise.all(membersNeedingApples.map(async (member) => {
+            if (userFrequency[member.id] >= 2) {
+                await member.roles.add(bothRoles);
+                usersWithGreenAppleAdded.add(member.id);
+                usersWithRedAppleAdded.add(member.id);
+            } else if (!member.roles.cache.has(greenAppleRoleId)) {
+                await member.roles.add([greenAppleRoleId]);
+                usersWithGreenAppleAdded.add(member.id);
+            } else {
+                await member.roles.add([redAppleRoleId]);
+                usersWithRedAppleAdded.add(member.id);
+            }
+            await removeRollCallRole(member);
+            await addActiveStudentRole(member);
+        }));
+    };
+
+    const addOnlyRedAppleRole = async () => {
+        if (membersNeedingOnlyRedApples.size === 0) {
+            return;
+        }
+        const members = Array.from(membersNeedingOnlyRedApples.values());
+        await Promise.all(members.map(async (member) => {
+            // must run after addAppleRoles to ensure cache has been updated
+            if (!member.roles.cache.has(redAppleRoleId)) {
```

Comment:
can user get the 🍎 if they dont have 🍏?

Reply:
yes! the helper logbooks only get a red apple. even if they have don't a green apple.